File: frontend/src/types/transport.ts
export interface VehicleLayoutCellSchema {
  row: number;
  col: number;
  type: string;
  label?: string | null;
  selectable?: boolean;
  blocked?: boolean;
  meta?: Record<string, unknown> | null;
}

export interface VehicleDeckSchema {
  key: string;
  label: string;
  rows: number;
  cols: number;
  cells: VehicleLayoutCellSchema[];
}

export interface VehicleLayoutSchema {
  rows?: number | null;
  cols?: number | null;
  cells: VehicleLayoutCellSchema[];
  decks?: VehicleDeckSchema[] | null;
}

export interface VehicleLayoutSummary {
  id: number;
  name: string;
  vehicle_type: string;
  slug?: string | null;
  seat_count: number;
  row_count: number;
  column_count: number;
  thumbnail_url?: string | null;
  is_active: boolean;
  created_at: string;
  updated_at?: string | null;
  deck_mode: "single" | "double";
  deck_summaries: VehicleDeckSummary[];
}

export interface VehicleLayoutDetail extends VehicleLayoutSummary {
  layout_schema: VehicleLayoutSchema;
}

export interface VehicleDeckSummary {
  key: string;
  label: string;
  seat_count: number;
  rows: number;
  cols: number;
}

export interface VehicleLayoutListResponse {
  items: VehicleLayoutSummary[];
}

export interface VehicleLayoutPayload {
  name: string;
  vehicle_type: string;
  slug?: string | null;
  seat_count?: number | null;
  row_count?: number | null;
  column_count?: number | null;
  thumbnail_url?: string | null;
  is_active: boolean;
  layout_schema: VehicleLayoutSchema;
}

export interface VehiclePayload {
  name: string;
  plate?: string | null;
  photo_url?: string | null;
  partner_name?: string | null;
  layout_id?: number | null;
  is_active: boolean;
}

export interface VehicleOut extends VehiclePayload {
  id: number;
  created_at: string;
  updated_at?: string | null;
}

export interface VehicleListResponse {
  items: VehicleOut[];
}

export interface SeatPostSignatureStatus {
  sale_id: number;
  contract_id?: number | null;
  product_id: number;
  signature_status?: string | null;
  is_road_trip: boolean;
  has_layout: boolean;
  seats_generated: boolean;
  payment_confirmed: boolean;
  passengers_completed: boolean;
  can_select_seats: boolean;
  seats_selected: boolean;
  message?: string | null;
}

export type TripVehicleStatus = "inactive" | "active" | "full";

export interface TripVehicleSummary {
  id: number;
  product_id: number;
  display_name?: string | null;
  vehicle_id?: number | null;
  layout_id?: number | null;
  capacity: number;
  occupied_seats: number;
  available_seats: number;
  order_index: number;
  status: TripVehicleStatus;
  is_active: boolean;
  auto_activate_next: boolean;
  vehicle?: VehicleOut | null;
  layout?: VehicleLayoutSummary | null;
  seats_generated: boolean;
  created_at: string;
  updated_at?: string | null;
}

export interface TripVehicleInput {
  id?: number | null;
  vehicle_id?: number | null;
  layout_id: number;
  display_name?: string | null;
  capacity: number;
  order_index: number;
  status?: TripVehicleStatus | null;
  is_active: boolean;
  auto_activate_next: boolean;
}

export interface TripSeatOut {
  id: number;
  trip_vehicle_id: number;
  seat_number: string;
  seat_label?: string | null;
  seat_type: string;
  row_index: number;
  col_index: number;
  is_selectable: boolean;
  is_blocked: boolean;
  is_occupied: boolean;
  occupied_by_passenger_id?: number | null;
  occupied_by_passenger_name?: string | null;
  occupied_by_sale_id?: number | null;
  occupied_by_current_sale: boolean;
  deck_key?: string | null;
  deck_label?: string | null;
  deck_order?: number | null;
}

export interface PassengerSeatSummary {
  id: number;
  sale_id: number;
  name: string;
  seat_id?: number | null;
  seat_number?: string | null;
  seat_label?: string | null;
  status: "unassigned" | "assigned";
}

export interface SeatMapStats {
  total_seats: number;
  available_seats: number;
  occupied_seats: number;
  blocked_seats: number;
  assigned_passengers: number;
  pending_passengers: number;
}

export interface TripTransportConfigOut {
  product_id: number;
  is_road_trip: boolean;
  layout?: VehicleLayoutSummary | null;
  vehicle?: VehicleOut | null;
  boarding_notes?: string | null;
  seats_generated: boolean;
  seat_count: number;
  updated_at?: string | null;
  vehicles: TripVehicleSummary[];
}

export interface TripTransportConfigPayload {
  is_road_trip: boolean;
  layout_id?: number | null;
  vehicle_id?: number | null;
  boarding_notes?: string | null;
  regenerate_layout?: boolean;
  vehicles?: TripVehicleInput[] | null;
}

export interface SeatMapContext {
  product_id: number;
  product_name: string;
  product_public_id?: string | null;
  trip_date?: string | null;
  trip_vehicle?: TripVehicleSummary | null;
  vehicles: TripVehicleSummary[];
  layout?: VehicleLayoutDetail | null;
  seats: TripSeatOut[];
  passengers: PassengerSeatSummary[];
  stats: SeatMapStats;
  boarding_notes?: string | null;
  can_assign: boolean;
}

export interface SeatSelectionContext extends SeatMapContext {
  sale_id: number;
  sale_reference?: string | null;
  can_submit: boolean;
  message?: string | null;
  preference_notice?: string | null;
}

export interface SeatAssignmentPayload {
  passenger_id: number;
  seat_id: number;
}

export interface SeatAdminAssignmentPayload extends SeatAssignmentPayload {
  notes?: string | null;
  assignment_status?: string | null;
}

export interface SeatBlockPayload {
  seat_id: number;
  is_blocked: boolean;
  reason?: string | null;
}

export interface SeatHistoryEntry {
  id: number;
  passenger_id?: number | null;
  passenger_name?: string | null;
  sale_id?: number | null;
  old_seat_label?: string | null;
  new_seat_label?: string | null;
  changed_by_role: string;
  changed_by_user_id?: number | null;
  reason?: string | null;
  created_at: string;
}

export interface SeatHistoryResponse {
  items: SeatHistoryEntry[];
  has_more: boolean;
}

File: backend/app/schemas/transport.py
from __future__ import annotations

from datetime import datetime, date
from typing import Any, Literal

from pydantic import BaseModel, Field, validator


class VehicleLayoutCellSchema(BaseModel):
    row: int
    col: int
    type: str
    label: str | None = None
    selectable: bool = True
    blocked: bool = False
    meta: dict[str, Any] | None = None


class VehicleDeckSchema(BaseModel):
    key: str
    label: str
    rows: int = Field(ge=1)
    cols: int = Field(ge=1)
    cells: list[VehicleLayoutCellSchema] = Field(default_factory=list)

    @validator("cells", each_item=True)
    def _validate_deck_cell(cls, cell: VehicleLayoutCellSchema, values: dict[str, Any]) -> VehicleLayoutCellSchema:
        rows = values.get("rows")
        cols = values.get("cols")
        if rows is not None and (cell.row < 0 or cell.row >= rows):
            raise ValueError("Cell row outside deck bounds")
        if cols is not None and (cell.col < 0 or cell.col >= cols):
            raise ValueError("Cell column outside deck bounds")
        return cell


class VehicleLayoutSchema(BaseModel):
    rows: int | None = Field(default=None, ge=1)
    cols: int | None = Field(default=None, ge=1)
    cells: list[VehicleLayoutCellSchema] = Field(default_factory=list)
    decks: list[VehicleDeckSchema] | None = None

    @validator("cells", each_item=True)
    def _validate_single_deck_cell(cls, cell: VehicleLayoutCellSchema, values: dict[str, Any]) -> VehicleLayoutCellSchema:
        if values.get("decks"):
            return cell
        rows = values.get("rows")
        cols = values.get("cols")
        if rows is None or cols is None:
            raise ValueError("rows and cols are required for single deck layouts")
        if cell.row < 0 or cell.row >= rows:
            raise ValueError("Cell row outside layout bounds")
        if cell.col < 0 or cell.col >= cols:
            raise ValueError("Cell column outside layout bounds")
        return cell


class VehicleLayoutPayload(BaseModel):
    name: str
    vehicle_type: str = "custom"
    slug: str | None = None
    seat_count: int | None = None
    row_count: int | None = None
    column_count: int | None = None
    thumbnail_url: str | None = None
    is_active: bool = True
    layout_schema: VehicleLayoutSchema


class VehicleLayoutSummary(BaseModel):
    id: int
    name: str
    vehicle_type: str
    slug: str | None = None
    seat_count: int
    row_count: int
    column_count: int
    thumbnail_url: str | None = None
    is_active: bool
    created_at: datetime
    updated_at: datetime | None = None
    deck_mode: Literal["single", "double"] = "single"
    deck_summaries: list["VehicleDeckSummary"] = Field(default_factory=list)


class VehicleLayoutDetail(VehicleLayoutSummary):
    layout_schema: VehicleLayoutSchema


class VehicleLayoutListResponse(BaseModel):
    items: list[VehicleLayoutSummary]


class VehiclePayload(BaseModel):
    name: str
    plate: str | None = None
    photo_url: str | None = None
    partner_name: str | None = None
    layout_id: int | None = None
    is_active: bool = True


class VehicleOut(VehiclePayload):
    id: int
    created_at: datetime
    updated_at: datetime | None = None


class VehicleListResponse(BaseModel):
    items: list[VehicleOut]


class SeatPostSignatureStatus(BaseModel):
    sale_id: int
    contract_id: int | None = None
    product_id: int
    signature_status: str | None = None
    is_road_trip: bool
    has_layout: bool
    seats_generated: bool
    payment_confirmed: bool
    passengers_completed: bool
    can_select_seats: bool
    seats_selected: bool
    message: str | None = None


class SaleSeatSelectionProduct(BaseModel):
    product_id: int
    product_public_id: str
    product_name: str


class SaleSeatSelectionProductsResponse(BaseModel):
    sale_id: int
    items: list[SaleSeatSelectionProduct] = Field(default_factory=list)


class TripVehiclePayload(BaseModel):
    id: int | None = None
    vehicle_id: int | None = None
    layout_id: int
    display_name: str | None = None
    capacity: int = Field(gt=0)
    order_index: int = Field(ge=1)
    status: Literal["inactive", "active", "full"] | None = None
    is_active: bool = True
    auto_activate_next: bool = True


class TripTransportConfigPayload(BaseModel):
    is_road_trip: bool
    layout_id: int | None = None
    vehicle_id: int | None = None
    boarding_notes: str | None = None
    regenerate_layout: bool = False
    vehicles: list[TripVehiclePayload] | None = None


class TripTransportConfigOut(BaseModel):
    product_id: int
    is_road_trip: bool
    layout: VehicleLayoutSummary | None = None
    vehicle: VehicleOut | None = None
    boarding_notes: str | None = None
    seats_generated: bool
    seat_count: int
    updated_at: datetime | None = None
    vehicles: list[TripVehicleOut] = Field(default_factory=list)


class TripSeatOut(BaseModel):
    id: int
    trip_vehicle_id: int
    seat_number: str
    seat_label: str | None = None
    seat_type: str
    row_index: int
    col_index: int
    is_selectable: bool
    is_blocked: bool
    is_occupied: bool
    occupied_by_passenger_id: int | None = None
    occupied_by_passenger_name: str | None = None
    occupied_by_sale_id: int | None = None
    occupied_by_current_sale: bool = False
    deck_key: str | None = None
    deck_label: str | None = None
    deck_order: int | None = None


class PassengerSeatSummary(BaseModel):
    id: int
    sale_id: int
    name: str
    seat_id: int | None = None
    seat_number: str | None = None
    seat_label: str | None = None
    status: Literal["unassigned", "assigned"]


class SeatMapStats(BaseModel):
    total_seats: int
    available_seats: int
    occupied_seats: int
    blocked_seats: int
    assigned_passengers: int
    pending_passengers: int


class SeatMapContext(BaseModel):
    product_id: int
    product_name: str
    product_public_id: str | None = None
    trip_date: date | None = None
    departure_instance_id: int | None = None
    departure_id: int | None = None
    departure_date: date | None = None
    departure_time: str | None = None
    trip_vehicle: TripVehicleOut | None = None
    vehicles: list[TripVehicleOut] = Field(default_factory=list)
    layout: VehicleLayoutDetail | None = None
    seats: list[TripSeatOut]
    passengers: list[PassengerSeatSummary]
    stats: SeatMapStats
    boarding_notes: str | None = None
    can_assign: bool = True


class SeatSelectionContext(SeatMapContext):
    sale_id: int
    sale_reference: str | None = None
    can_submit: bool = True
    message: str | None = None
    preference_notice: str | None = None


class SeatAssignmentPayload(BaseModel):
    passenger_id: int
    seat_id: int


class SeatAdminAssignmentPayload(SeatAssignmentPayload):
    notes: str | None = None
    assignment_status: str | None = None


class SeatBlockPayload(BaseModel):
    seat_id: int
    is_blocked: bool
    reason: str | None = None


class SeatHistoryEntry(BaseModel):
    id: int
    passenger_id: int | None = None
    passenger_name: str | None = None
    sale_id: int | None = None
    old_seat_label: str | None = None
    new_seat_label: str | None = None
    changed_by_role: str
    changed_by_user_id: int | None = None
    reason: str | None = None
    created_at: datetime


class SeatHistoryResponse(BaseModel):
    items: list[SeatHistoryEntry]
    has_more: bool


class VehicleDeckSummary(BaseModel):
    key: str
    label: str
    seat_count: int
    rows: int
    cols: int


class TripVehicleOut(BaseModel):
    id: int
    product_id: int
    display_name: str | None = None
    vehicle_id: int | None = None
    layout_id: int | None = None
    capacity: int
    occupied_seats: int
    available_seats: int
    order_index: int
    status: Literal["inactive", "active", "full"]
    is_active: bool
    auto_activate_next: bool
    vehicle: VehicleOut | None = None
    layout: VehicleLayoutSummary | None = None
    seats_generated: bool
    created_at: datetime
    updated_at: datetime | None = None


class TripVehicleListResponse(BaseModel):
    items: list[TripVehicleOut]


VehicleLayoutSummary.model_rebuild()
VehicleLayoutDetail.model_rebuild()
TripTransportConfigOut.model_rebuild()
TripVehicleOut.model_rebuild()
